cloud-storage/cloud_storage.c

```diff
@@ -0,0 +1,909 @@
+
+      struct
+      {
+         folder_type_t folder_type;
+         char *local_file;
+      } upload;
+
+      struct
+      {
+         folder_type_t folder_type;
+         cloud_storage_item_t *remote_file;
+      } download;
+   } operation_parameters;
+
+   struct _operation_queue_item *next;
+};
+
+static bool _shutdown = true;
+static cloud_storage_provider_t *_provider;
+
+static char _game_states_config_value[8192] = "";
+
+static bool _need_sync_runtime_logs = false;
+
+cloud_storage_item_t *_game_states = NULL;
+
+static struct _operation_queue_item *_operation_queue = NULL;
+static slock_t *_operation_queue_mutex;
+
+static scond_t *_operation_condition;
+static sthread_t *_operation_thread;
+
+static void _sync_files_execute(folder_type_t folder_type);
+
+static void _upload_file_execute(folder_type_t folder_type, char *file_name);
+
+void cloud_storage_item_free(cloud_storage_item_t *item)
+{
+   cloud_storage_item_t *next;
+
+   while (item)
+   {
+      next = item->next;
+
+      free(item->id);
+      free(item->name);
+
+      if (item->item_type == CLOUD_STORAGE_FILE)
+      {
+         if (item->type_data.file.hash_value)
+         {
+            free(item->type_data.file.hash_value);
+         }
+         if (item->type_data.file.download_url)
+         {
+            free(item->type_data.file.download_url);
+         }
+      }
+
+      if (item->item_type == CLOUD_STORAGE_FOLDER && item->type_data.folder.children)
+      {
+         cloud_storage_item_free(item->type_data.folder.children);
+      }
+
+      free(item);
+
+      item = next;
+   }
+}
+
+static void _free_operation_queue(struct _operation_queue_item *queue)
+{
+   struct _operation_queue_item *next_item;
+
+   while (queue)
+   {
+      switch (queue->operation)
+      {
+         case UPLOAD_FILE:
+            if (queue->operation_parameters.upload.local_file)
+            {
+               free(queue->operation_parameters.upload.local_file);
+            }
+            break;
+         case DOWNLOAD_FILE:
+            if (queue->operation_parameters.download.remote_file)
+            {
+               cloud_storage_item_free(queue->operation_parameters.download.remote_file);
+            }
+            break;
+         default:
+            break;
+      }
+
+      next_item = queue->next;
+      free(queue);
+      queue = next_item;
+   }
+}
+
+static void _operation_thread_loop(void *user_data)
+{
+   while (!_shutdown)
+   {
+      struct _operation_queue_item *task;
+
+      slock_lock(_operation_queue_mutex);
+
+      scond_wait_timeout(_operation_condition, _operation_queue_mutex, 500);
+      if (_shutdown)
+      {
+         slock_unlock(_operation_queue_mutex);
+         continue;
+      }
+
+      task = _operation_queue;
+      if (_operation_queue)
+      {
+         _operation_queue = NULL;
+      }
+
+      slock_unlock(_operation_queue_mutex);
+
+      while (task)
+      {
+         struct _operation_queue_item *next_task;
```

Comment:
I don't think this queue code is properly implemented, but I dont have more proof than reading it.
In the queue add item routine, it modifies the last element next field to queue a new operation. Here, the task next is accessed without grabbing the lock, that sounds like trouble to me.
I'd recommend you to get rid of this while loop and simply "pop" the front element before (in the area where you grab the lock). That way _shutdown is also evaluated, allowing to early exit the queue if needed.

Reply:
On line 164, I get the lock. I then assign the variable **task** on line 173 to be the operation queue and set the operation queue to NULL. Finally the lock is released on line 179.

The basic idea is to take all tasks from the queue and then process them.

This ties in to your comment above. Since it is checking for tasks every 0.5 seconds, I am trying to take all tasks. If I can get the condition working as expected, then this can be changed to only work on one task.

Reply:
This has been changed to only work on the first item in the queue (which is the oldest item).